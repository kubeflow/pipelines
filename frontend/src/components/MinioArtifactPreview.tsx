import * as React from 'react';
import { StorageService } from '../lib/WorkflowParser';
import { S3Artifact } from '../../third_party/argo-ui/argo_template';
import { generateArtifactUrl, isS3Endpoint } from '../lib/Utils';
import { Apis } from '../lib/Apis';


/**
 * Check if a javascript object is an argo S3Artifact object.
 * @param value Any javascript object.
 */
export function isS3Artifact(value: any): value is S3Artifact {
  return value && value.key && value.bucket && value.endpoint;
}

/**
 * A component that renders a preview to an artifact with a link to the full content.
 */
const MinioArtifactPreview: React.FC<{artifact: S3Artifact}> = ({artifact = {}}) => {
  const { key, bucket, endpoint } = artifact;
  if (!key || !bucket) {
    return null;
  }

  const peek = 100;
  const encodedKey = encodeURIComponent(key);
  const source = isS3Endpoint(endpoint) ? StorageService.S3 : StorageService.MINIO;
  const linkText = `${source.toString()}://${bucket}/${encodedKey}`;
  const url = generateArtifactUrl(source, bucket, encodedKey);
  const [content, setContent] = React.useState('');

  React.useEffect(() => {
    let cancelled = false;
    Apis.readFile({ source, bucket, key }, peek + 1).then(
      data => !cancelled && setContent(data.length > peek ? `${data.slice(0, peek)} ...` : data),
      error => console.error(error),
    );
    return () => {
      cancelled = true;
    };
  }, [source, bucket, key]);

  // Opens in new window safely
  return (
    <div>
      <a href={url} target={'_blank'} rel={'noreferrer noopener'} title={linkText}>
        {linkText}
      </a>
      <div>
        <small><pre>{content}</pre></small>
      </div>
    </div>
  );
};

export default MinioArtifactPreview;
