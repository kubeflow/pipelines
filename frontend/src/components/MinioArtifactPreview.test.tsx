/*
 * Copyright 2019-2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { mount } from 'enzyme'
import MinioArtifactPreview from './MinioArtifactPreview';

describe('MinioArtifactPreview', () => {
  it('handles undefined artifact', () => {
    expect(<MinioArtifactPreview artifact={undefined as any} />).toMatchSnapshot();
  });

  it('handles null artifact', () => {
    expect(<MinioArtifactPreview artifact={null as any} />).toMatchSnapshot();
  });

  it('handles empty artifact', () => {
    expect(<MinioArtifactPreview artifact={{} as any} />).toMatchSnapshot();
  });

  it('handles invalid artifact: no bucket', () => {
    const s3artifact = {
      accessKeySecret: { key: 'accesskey', optional: false, name: 'minio' },
      bucket: '',
      endpoint: 'minio.kubeflow',
      key: 'bar',
      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
    };
    expect(<MinioArtifactPreview artifact={s3artifact as any} />).toMatchSnapshot();
  });

  it('handles invalid artifact: no key', () => {
    const s3artifact = {
      accessKeySecret: { key: 'accesskey', optional: false, name: 'minio' },
      bucket: 'foo',
      endpoint: 'minio.kubeflow',
      key: '',
      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
    };
    expect(<MinioArtifactPreview artifact={s3artifact as any} />).toMatchSnapshot();
  });

  it('handles s3 artifact', () => {
    const s3artifact = {
      accessKeySecret: { key: 'accesskey', optional: false, name: 'minio' },
      bucket: 'foo',
      endpoint: 's3.amazonaws.com',
      key: 'bar',
      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
    };
    const wrapper = mount(<MinioArtifactPreview artifact={s3artifact} />)
    expect(wrapper).toMatchSnapshot()
  });

  it('handles minio artifact', () => {
    const minioartifact = {
      accessKeySecret: { key: 'accesskey', optional: false, name: 'minio' },
      bucket: 'foo',
      endpoint: 'minio.kubeflow',
      key: 'bar',
      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
    };
    const wrapper = mount(<MinioArtifactPreview artifact={minioartifact} />)
    expect(wrapper).toMatchSnapshot()
  });
});
