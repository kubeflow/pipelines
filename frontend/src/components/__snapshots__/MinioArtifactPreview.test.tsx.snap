// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinioArtifactPreview handles empty artifact 1`] = `
<MinioArtifactPreview
  artifact={Object {}}
/>
`;

exports[`MinioArtifactPreview handles invalid artifact: no bucket 1`] = `
<MinioArtifactPreview
  artifact={
    Object {
      "accessKeySecret": Object {
        "key": "accesskey",
        "name": "minio",
        "optional": false,
      },
      "bucket": "",
      "endpoint": "minio.kubeflow",
      "key": "bar",
      "secretKeySecret": Object {
        "key": "secretkey",
        "name": "minio",
        "optional": false,
      },
    }
  }
/>
`;

exports[`MinioArtifactPreview handles invalid artifact: no key 1`] = `
<MinioArtifactPreview
  artifact={
    Object {
      "accessKeySecret": Object {
        "key": "accesskey",
        "name": "minio",
        "optional": false,
      },
      "bucket": "foo",
      "endpoint": "minio.kubeflow",
      "key": "",
      "secretKeySecret": Object {
        "key": "secretkey",
        "name": "minio",
        "optional": false,
      },
    }
  }
/>
`;

exports[`MinioArtifactPreview handles minio artifact 1`] = `
<MinioArtifactPreview
  artifact={
    Object {
      "accessKeySecret": Object {
        "key": "accesskey",
        "name": "minio",
        "optional": false,
      },
      "bucket": "foo",
      "endpoint": "minio.kubeflow",
      "key": "bar",
      "secretKeySecret": Object {
        "key": "secretkey",
        "name": "minio",
        "optional": false,
      },
    }
  }
>
  <div>
    <a
      href="artifacts/get?source=minio&bucket=foo&key=bar"
      rel="noreferrer noopener"
      target="_blank"
      title="minio://foo/bar"
    >
      minio://foo/bar
    </a>
    <div>
      <small>
        <pre />
      </small>
    </div>
  </div>
</MinioArtifactPreview>
`;

exports[`MinioArtifactPreview handles null artifact 1`] = `
<MinioArtifactPreview
  artifact={null}
/>
`;

exports[`MinioArtifactPreview handles s3 artifact 1`] = `
<MinioArtifactPreview
  artifact={
    Object {
      "accessKeySecret": Object {
        "key": "accesskey",
        "name": "minio",
        "optional": false,
      },
      "bucket": "foo",
      "endpoint": "s3.amazonaws.com",
      "key": "bar",
      "secretKeySecret": Object {
        "key": "secretkey",
        "name": "minio",
        "optional": false,
      },
    }
  }
>
  <div>
    <a
      href="artifacts/get?source=s3&bucket=foo&key=bar"
      rel="noreferrer noopener"
      target="_blank"
      title="s3://foo/bar"
    >
      s3://foo/bar
    </a>
    <div>
      <small>
        <pre />
      </small>
    </div>
  </div>
</MinioArtifactPreview>
`;

exports[`MinioArtifactPreview handles undefined artifact 1`] = `
<MinioArtifactPreview
  artifact={undefined}
/>
`;
